src/main/java/engineering/swat/watch/impl/JDKRecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,317 @@
+package engineering.swat.watch.impl;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import javax.management.RuntimeErrorException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import engineering.swat.watch.WatchEvent;
+
+public class JDKRecursiveDirectoryWatcher implements Closeable {
+    private final Logger logger = LogManager.getLogger();
+    private final Path root;
+    private final Executor exec;
+    private final Consumer<WatchEvent> eventHandler;
+    private final ConcurrentMap<Path, JDKDirectoryWatcher> activeWatches = new ConcurrentHashMap<>();
+
+    public JDKRecursiveDirectoryWatcher(Path directory, Executor exec, Consumer<WatchEvent> eventHandler) {
+        this.root = directory;
+        this.exec = exec;
+        this.eventHandler = eventHandler;
+    }
+
+    public void start() throws IOException {
+        try {
+            logger.debug("Starting recursive watch for: {}", root);
+            registerInitialWatches(root);
+        } catch (IOException e) {
+            throw new IOException("Could not register directory watcher for: " + root, e);
+        }
+    }
+
+    private void processEvents(WatchEvent ev) {
+        logger.trace("Forwarding event: {}", ev);
+        eventHandler.accept(ev);
+        logger.trace("Unwrapping event: {}", ev);
+        try {
+            switch (ev.getKind()) {
+                case CREATED: handleCreate(ev); break;
+                case DELETED: handleDeleteDirectory(ev); break;
+                case OVERFLOW: handleOverflow(ev); break;
+                case MODIFIED: break;
+            }
+        } finally {
```

Comment:
Nit: Remove this `try`/`finally`?

Reply:
ah yes, it's a leftover from when the extra event handling was done synchronously.